blog/_src/posts/2016-05-18-gradual-typing-across-the-spectrum.md

```diff
@@ -0,0 +1,181 @@
+Typed Racket, which would allow correct-by-construction programming using
+a Nuprl-like proof system (he had a very cool GUI proof assistant demo in
+his slides!).
+
+---
+
+One of the key practical concerns about gradual typing is its performance overhead.
+It's a concern because in order to ensure type safety, a gradually-typed language
+implementation needs to install dynamic checks between the typed and untyped
+parts of a program. This catches any inconsistencies between the typed interfaces
+and how the untyped code may call into them.
+
+Ben Greenman gave an upbeat talk that set the stage for this topic, pointing out
+some key lessons that we've learned about performance from building Typed Racket.
+The main idea he presented (also the topic of our [POPL 2016 paper](http://www.ccs.neu.edu/racket/pubs/popl16-tfgnvf.pdf))
+is that to evaluate a gradual type system, you want to explore different ways
+of adding types to a program and see how much it costs. This evaluation effort
+started with Typed Racket, but he and Zeina Migeed are working on expanding it
+to Reticulated Python.
+
+From IU, Andre Kuhlenschmidt and Deyaaeldeen Almahallawi are exploring how
+ahead-of-time (AOT) compilation strategies could help reduce the cost of gradual
+typing. In particular, they are working on implementing [Schml](https://github.com/deyaaeldeen/Schml):
+a compiler from the gradually-typed lambda calculus to C.
+
+In addition to AOT compilation, the folks at IU are exploring tracing JIT compilation
+as a means to make gradual typing faster. More specifically,
+Spenser Bauman talked about Pycket, an alternative implementation of Racket that
+uses RPython/PyPy to dramatically lower the overhead of gradual typing (also see the
+[recording](https://www.youtube.com/watch?v=GOfIY8NHAqg) of Spenser's
+talk on the topic at RacketCon and his [ICFP paper](http://homes.soic.indiana.edu/samth/pycket-draft.pdf)).
+
+---
+
+On the usability side, both Shriram Krishnamurthi and Ambrose Bonnaire-Sergeant
+made observations on what it takes to get gradual typing in the hands of real
+software developers.
+
+Shriram approached the topic from the angle of CS education, which is the focus
+of the [Pyret](http://www.pyret.org) language, and shared what the Brown language
+group is working on. While Pyret doesn't exactly fit the mold of gradual typing,
+it's a close cousin since it's a dynamically-typed language that explicitly takes
+design cues from the best parts of statically-typed languages. That approach
+lets CS beginners think in terms of types (the approach spearheaded by
+[HtDP](http://www.ccs.neu.edu/home/matthias/HtDP2e/index.html) and [Bootstrap](http://www.bootstrapworld.org/))
+without having to battle a typechecker from the start.
+
+For professional software developers, a major concern with gradual typing is that
+writing type annotations is a tedious and time intensive task. Ambrose, who
```

Comment:
Maybe "may be a tedious and time intensive talk"? The phrase, unqualified, sounds possibly wrong to me, although it may be very true in the specific systems used for gradual typing today or given the state of type inference for them.